src/compositor.rs

```diff
@@ -977,16 +977,22 @@ impl<G: GraphicsBackend> FrameController<G> {
                         "App submitted a null texture and a normal texture in the same frame"
                     );
 
-                    if self.swapchain_data.as_ref().is_none() {
-                        // SAFETY: Technically SessionCreateInfo should be Copy anyway so this should be fine:
-                        // https://github.com/Ralith/openxrs/issues/183
-                        self.recreate_swapchain(session_data, unsafe { std::ptr::read(&new_info) });
-                    }
-                    let data = self.swapchain_data.as_ref().unwrap();
-                    if !is_usable_swapchain(&data.info, data.initial_format, &new_info) {
-                        info!("recreating swapchain (for {eye:?})");
+                    if self
+                        .swapchain_data
+                        .as_ref()
+                        .map(|data| {
+                            !is_usable_swapchain(&data.info, data.initial_format, &new_info)
+                        })
+                        .inspect(|not_usable| {
+                            if *not_usable {
+                                info!("recreating swapchain (for {eye:?})");
+                            }
+                        })
+                        .unwrap_or(true)
```

Comment:
I think it's clearer if you reverse the condition:
```suggestion
                    if !self
                        .swapchain_data
                        .as_ref()
                        .is_some_and(|data| {
                            is_usable_swapchain(&data.info, data.initial_format, &new_info)
                        })
                        .inspect(|usable| {
                            if !usable {
                                info!("recreating swapchain (for {eye:?})");
                            }
                        })
                        .unwrap_or(false)
```

Reply:
`is_some_and` returns a bool, so that won't work with `.inspect`.

Reply:
What about a simple `match`? I find that a lot more readable, even if it means having two branches that call `recreate_swapchain()`.

Reply:
I don't see how a match would be significantly more readable. I'd just move the recreating swapchain log inside of the if block, I don't think it's that important to super precise with the log.